### ISSUE TITLE
Remove "orphaned" nodes left after trimming the dangling edges.

### ISSUE THREAD

**Issue by larsop** [NibioOpenSource/pgtopo_update_sql#5]

**Comment by strk** [comment]
Generic "orphaned" node deleter, assuming no TopoGeometry is ever defined by a node:

delete from XX.node where node_id not in ( select start_node from XX.edge union select end_node from XX.edge );

**Comment by larsop** [comment]
For some layers we may have single points that is referred from a single topo objects, but not this layer.

But since we are trying a make generic update code, all clean up code should only clean up data that are created by the process it self.

**Comment by strk** [comment]
Genericity is the reason why I think you should not mix the cleaning
of primitive objects and TopoGeometry definitions.

The definition of a TopoGeometry is the final product of the toTopoGeom
function, but you can make no assumption on which elements in its
definition existed before the input geometry was converted.

A generic "cleaner" function could remove any "orphaned" primitive,
being a primitive object that's not used in the definition of any
TopoGeometry, starting from edges.

An edge is unused if all the following conditions are met:
- None of the faces on the edge side are referenced by a TopoGeometry
- The edge is not referenced by a TopoGeometry

After edges are removed, isolated nodes can be removed if:
- The node is not referenced by a TopoGeometry

**Comment by larsop** [comment]
Thanks for feedback it's nice to get discussion on this before we start the work. 

I see your arguments. I am not sure whats the best solution is.

Here are some reasons why I am unsure.
- If we have big layers with billions of points this generic clean up operation may take some time, but we can run this in the background at night so that may not be a problem.
- I though it should be possible to find out if there are added any new objects by checking the content in the area I work on before and after running  toTopoGeom function. I partly do this in some other cases to day, by creating temp tables which contains data before a topo functions area called. Then compare this with new data after the topo the topo functions are called. But this again makes the update process more complex and take more time.
- There was a new argument coming today and that was from the user interface side and that was that in some cases users may create non used topo elements for shorter a period of time (some minutes).

Since I am so unsure about how to this I just wait with this problem until  later, but I tend to agree with you about using a generic function in some way, if it was not for the user interface discussion we had today.

I will decide on how to this when we done work on the user interface.

**Comment by strk** [comment]
On Wed, Sep 09, 2015 at 02:51:38AM -0700, Lars Aksel Opsahl wrote:

> - If we have big layers with billions of points this generic clean up operation may take some time, but we can run this in the background at night so that may not be a problem.

Yes, it can be an idle time operation.
And it doesn't need to be complete (can stop at any time).

> - I though it should be possible to find out if there are added any new objects by checking the content in the area I work on before and after running  toTopoGeom function. I partly do this in some other cases to day, by creating temp tables which contains data before a topo functions area called. Then compare this with new data after the topo the topo functions are called. But this again makes the update process more complex and take more time.

You could limit the operations of the GC function on a given extent,
and use the extent of the input geometry (no need to be complete).

> - There was a new argument coming today and that was from the user interface side and that was that in some cases users may create non used topo elements for shorter a period of time (some minutes).

This is a good point. Could the UI side create temporary TopoGeometry
objects to go with the temporary topo elements ?

**Comment by larsop** [comment]
Yes in way when may need "temporary" surface objects.

When the end user is updating the map, the user draws some new lines and the map is changed for the surface layer even if the user is drawing a non closed linestring.

We will probably not need a full topo object but just return the  topology.TopoElementArray_Agg for the faces generated so the client can draw this in the map. 

So before the operation is valid the user needs to set some surface attributes and then it's nice if he can see can the changes in the map before he sets the attributes and maybe also select the surface(s) generated.

The way postgis topology is built up makes this way of working quite easy.

**Comment by strk** [comment]
The "full" topo object would be needed to give an hint to the
primitive elements cleaner that those primitive elements are in use.

You can deploy the TopoGeometry object in a temporary topology layer
built for the purpose and run clearTopoGeom on the objects when they
are not needed anymore.

**Comment by larsop** [comment]
Yes to have temp topoobjects is a good idea and I agree that if we run this as generic cleaner we need a "full" topo object.

If the orphaned objects are not casing problem for the UI  operations we just implement this cleaner after we have a system up running. 

If we have a none generic cleaner we can do this in different ways.
